src/julia_internal.h

```diff
@@ -186,79 +192,61 @@ static const int jl_gc_sizeclasses[JL_GC_N_POOLS] = {
+    // is a multiple of the alignment. As an example an allocation with
+    // sz of 48 and an alignment of 32 will need to be in pool of size 64.
+    // An alignment of 0 or 1 means unaligned and we can use sz directly.
+    if (alignment != 0 && alignment != 1 && alignment != sz)
+        sz = ((sz / alignment) + 1) * alignment;
+    return sz;
 }
-JL_DLLEXPORT int jl_alignment(size_t sz);
 
-STATIC_INLINE int JL_CONST_FUNC jl_gc_szclass(size_t sz)
+// Use jl_gc_alignsz to obtain the right sz.
+STATIC_INLINE int JL_CONST_FUNC jl_gc_szclass(size_t sz, size_t alignment)
 {
-#ifdef _P64
-    if (sz <=    8)
-        return 0;
-    const int N = 0;
-#elif defined(_CPU_ARM_) || defined(_CPU_PPC_)
-    if (sz <=    8)
-        return (sz + 3) / 4 - 1;
-    const int N = 1;
-#else
-    if (sz <=   12)
-        return (sz + 3) / 4 - 1;
-    const int N = 2;
-#endif
-    if (sz <=  256)
-        return (sz + 15) / 16 + N;
-    if (sz <=  496)
-        return 16 - 16376 / 4 / LLT_ALIGN(sz, 16 * 4) + 16 + N;
-    if (sz <= 1008)
-        return 16 - 16376 / 2 / LLT_ALIGN(sz, 16 * 2) + 24 + N;
-    return     16 - 16376 / 1 / LLT_ALIGN(sz, 16 * 1) + 32 + N;
+    size_t klass = 0;
+    while (klass < JL_GC_N_POOLS) {
+        size_t osize = jl_gc_sizeclasses[klass];
+        if (sz <= osize) {
+            if (alignment == 1 || alignment == 0 || osize % alignment == 0)
+                return klass;
+        }
+        klass++;
+    }
+    return -1;
 }
 
 #ifdef __GNUC__
 #  define jl_is_constexpr(e) __builtin_constant_p(e)
 #else
 #  define jl_is_constexpr(e) (0)
 #endif
-#define JL_SMALL_BYTE_ALIGNMENT 16
-#define JL_CACHE_BYTE_ALIGNMENT 64
-// JL_HEAP_ALIGNMENT is the maximum alignment that the GC can provide
-#define JL_HEAP_ALIGNMENT JL_SMALL_BYTE_ALIGNMENT
-#define GC_MAX_SZCLASS (2032-sizeof(void*))
 
-STATIC_INLINE jl_value_t *jl_gc_alloc_(jl_ptls_t ptls, size_t sz, void *ty)
+STATIC_INLINE jl_value_t *jl_gc_alloc_(jl_ptls_t ptls, size_t sz, size_t alignment, void *ty)
 {
     const size_t allocsz = sz + sizeof(jl_taggedvalue_t);
     if (allocsz < sz) // overflow in adding offs, size was "negative"
         jl_throw(jl_memory_exception);
+    const size_t alignsz = jl_gc_alignsz(allocsz, alignment);
+    assert(alignsz >= allocsz);
+    const int klass = jl_gc_szclass(alignsz, alignment);
     jl_value_t *v;
-    if (allocsz <= GC_MAX_SZCLASS + sizeof(jl_taggedvalue_t)) {
-        int pool_id = jl_gc_szclass(allocsz);
+    if (klass != -1 && alignsz <= GC_MAX_SZCLASS + sizeof(jl_taggedvalue_t)) {
+        int pool_id = klass;
         jl_gc_pool_t *p = &ptls->heap.norm_pools[pool_id];
         int osize;
-        if (jl_is_constexpr(allocsz)) {
+        if (jl_is_constexpr(alignsz)) {
             osize = jl_gc_sizeclasses[pool_id];
         }
         else {
             osize = p->osize;
         }
+        assert((size_t)osize >= alignment &&
+               (alignment == 0 || alignment == 1 || osize % alignment == 0));
```

Comment:
`alignment == 0 || (osize & (alignment - 1)) == 0` (if you want, also can add `&& (alignment & (alignment - 1) == 0)` to assert that alignment is a power-of-two)